Saule/Serialization/ResourceGraph.cs

```diff
@@ -0,0 +1,144 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+using Saule.Queries.Including;
+
+namespace Saule.Serialization
+{
+    internal class ResourceGraph
+    {
+        private readonly IDictionary<ResourceGraphNodeKey, ResourceGraphNode> _nodes;
+
+        public ResourceGraph(
+            object obj,
+            ApiResource resource,
+            ResourceGraphPathSet includePaths)
+        {
+            obj.ThrowIfNull(nameof(obj));
+            resource.ThrowIfNull(nameof(resource));
+            includePaths.ThrowIfNull(nameof(includePaths));
+
+            _nodes = new Dictionary<ResourceGraphNodeKey, ResourceGraphNode>();
+
+            Build(obj, resource, includePaths, 0);
+        }
+
+        public IEnumerable<ResourceGraphNode> DataNodes
+        {
+            get
+            {
+                return _nodes.Values.Where(v => v.GraphDepth == 0);
+            }
+        }
+
+        public IEnumerable<ResourceGraphNode> IncludedNodes
+        {
+            get
+            {
+                return _nodes.Values.Where(v => v.GraphDepth > 0);
+            }
+        }
+
+        private void Build(
+            object obj,
+            ApiResource resource,
+            ResourceGraphPathSet includePaths,
+            int depth)
+        {
+            if (obj == null)
+            {
+                // end of the line - obj will be a leaf node in our graph
+                return;
+            }
+
+            if (obj is IEnumerable)
```

Comment:
Ok perhaps I'm not completely clear on this. 

The expression tree is already evaluated and a response generated before the graph starts reflecting upon it. Is this what you meant @joukevandermaas?